### PR TITLE
Document and add tests to Faceted Search system

### PR DESCRIPTION
**Pull request by mickaelandrieu**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Added PHPDocs + minor refactoring + LOT OF unit tests
| Type?         | improvement
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| How to test?  | Facet search in Front office should work as before.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/10406)

### CODE REVIEW

File: src/Core/Product/Search/ProductSearchResult.php [hunk 1]
```diff
@@ -28,73 +28,136 @@
...
     public function setFacetCollection(FacetCollection $facetCollection)
     {
         $this->facetCollection = $facetCollection;
 
         return $this;
     }
 
+    /**
+     * @return mixed
+     */
```
Comment (jolelievre) [inline src/Core/Product/Search/ProductSearchResult.php:110]: `@return FacetCollection`

File: src/Core/Product/Search/SortOrder.php [hunk 2]
```diff
@@ -65,74 +131,107 @@ public function toArray()
...
+    public function setDirection($direction)
     {
-        $direction = strtolower($dir);
-        if (!in_array($direction, ['asc', 'desc', 'random'])) {
+        $formattedDirection = strtolower($direction);
+        if (!in_array($formattedDirection, ['asc', 'desc', 'random'])) {
             throw new Exception(sprintf(
                 'Invalid SortOrder direction `%s`. Expecting one of: `ASC`, `DESC`, or `RANDOM`.',
-                $dir
+                $direction
```
Comment (jolelievre) [inline src/Core/Product/Search/SortOrder.php]: possible to have a more specific Exception? `SearchException` ?

File: src/Core/Product/Search/URLFragmentSerializer.php [hunk 3]
```diff
@@ -78,6 +95,11 @@ private function unserializeListOfStrings($separator, $escape, $str)
         return $list;
     }
 
+    /**
+     * @param array $fragment
+     *
+     * @return string
+     */
```
Comment (jolelievre) [inline src/Core/Product/Search/URLFragmentSerializer.php]: public methods should be placed before private ones

File: src/Core/Product/Search/SortOrder.php [hunk 4]
```diff
@@ -65,74 +131,107 @@ public function toArray()
...
     public function getField()
     {
         return $this->field;
     }
 
-    public function setDirection($dir)
+    /**
+     * @param string $direction
+     *
+     * @return string
```
Comment (jolelievre) [inline src/Core/Product/Search/SortOrder.php:191]: what is the purpose of returning a string here (`$this->direction`) ? whereas all other setters return `$this` which more useful to cascade setters

Comment (mickaelandrieu) [reply src/Core/Product/Search/SortOrder.php:191]: I don't know but I can't break the public API :/

File: src/Core/Product/Search/SortOrder.php (hunk 2 above)
Comment (mickaelandrieu) [reply src/Core/Product/Search/SortOrder.php]: ~~I'm not supposed to change the public API => it's a BC~~ well it's better to have a specific exception than a hidden error. Ok for that.

**Comment by mickaelandrieu** [comment]
Comments addressed, please review :)

Thanks @jolelievre :)

**Comment by mickaelandrieu** [comment]
Tests are still failing but I can't figure out how to reproduce it locally, can someone take a look? @jolelievre maybe?

**Comment by jolelievre** [comment]
ouh I am far from being an expert on the E2E tests ^^ just to be sure did you rebase your branche with recent changes on these tests? but I think they are pended with the 1.7.5>develop PR

**Comment by mickaelandrieu** [comment]
yes, it's rebase from "16 hours ago" :+1:

**Review by eternoendless — CHANGES_REQUESTED** [review]
> Reviewed 7 of 24 files at r1, 9 of 18 files at r2.
> *[Reviewable](https://reviewable.io/reviews/prestashop/prestashop/10406)* status: all files reviewed, 7 unresolved discussions (waiting on @jolelievre and @mickaelandrieu)

___
*[src/Core/Product/Search/SortOrderFactory.php, line 34 at r2](https://reviewable.io/reviews/prestashop/prestashop/10406#-LYbTnUA961Vplo_ksKa:-LYbTnUA961Vplo_ksKb:b-uukar5) ([raw file](https://github.com/prestashop/prestashop/blob/d4d0bbceba86e6c86ce04716317634b94d1e4f0f/src/Core/Product/Search/SortOrderFactory.php#L34)):*
> ```PHP
>  * This class is not a factory but a provider of default Sort Orders.
>  *
>  * @deprecated since 1.7.5 and to be removed in 1.8: use SortOrdersCollection:getDefaults instead.
> ```

We'll need to update this to 1.7.6
___
*[src/Core/Product/Search/SortOrderFactory.php, line 44 at r2](https://reviewable.io/reviews/prestashop/prestashop/10406#-LYbTrdVEd2rOeZrJy1C:-LYbTrdVEd2rOeZrJy1D:bbqep6w) ([raw file](https://github.com/prestashop/prestashop/blob/d4d0bbceba86e6c86ce04716317634b94d1e4f0f/src/Core/Product/Search/SortOrderFactory.php#L44)):*
> ```PHP
> 
>     public function __construct(TranslatorInterface $translator)
>     {
> ```

We should issue a deprecation notice here.
___
*[src/Core/Product/Search/SortOrderFactory.php, line 55 at r2](https://reviewable.io/reviews/prestashop/prestashop/10406#-LYb_vA539iTUegmiL-V:-LYb_vA539iTUegmiL-W:b-6nup9z) ([raw file](https://github.com/prestashop/prestashop/blob/d4d0bbceba86e6c86ce04716317634b94d1e4f0f/src/Core/Product/Search/SortOrderFactory.php#L55)):*
> ```PHP
>     public function getDefaultSortOrders()
>     {
>         return [
> ```

You could use the replacement class here to avoid duplicate code.
___
*[tests/Unit/Core/Product/Search/FacetTest.php, line 148 at r2](https://reviewable.io/reviews/prestashop/prestashop/10406#-LYbUzA62Mc_vduEjlH7:-LYbUzA62Mc_vduEjlH8:bgdw2pn) ([raw file](https://github.com/prestashop/prestashop/blob/d4d0bbceba86e6c86ce04716317634b94d1e4f0f/tests/Unit/Core/Product/Search/FacetTest.php#L148)):*
> ```PHP
>         $this->assertInstanceOf(Facet::class, $this->facet->setProperty('product_name', 'Nice cupcake'));
>         $this->assertSame('Nice cupcake', $this->facet->getProperty('product_name'));
>     }
> ```

Not sure that there's value in testing getters and setters. There's an interesting discussion about it here: https://stackoverflow.com/questions/6197370/should-unit-tests-be-written-for-getter-and-setters/

**Comment by eternoendless** [comment]
I'm ok for merging all the comments as they are!

**Comment by mickaelandrieu** [comment]
> I'm ok for merging all the comments as they are!

I will update my pull request: thanks for your review @eternoendless :+1:

**Comment by mickaelandrieu** [comment]
> Not sure that there's value in testing getters and setters. There's an interesting discussion about it here: https://stackoverflow.com/questions/6197370/should-unit-tests-be-written-for-getter-and-setters/

You can't guess how many times a getter/setter is misused or introduce a bug :laughing:  Also, having to have to write tests for them force any developer to think about the existence of these getter/setter: are they required or even useful?

Finally, they are really easy and fast to create and to execute. And as SOLID requires that a class should have 1 and only 1 reason to change: these tests don't need to be maintained, am I right? :angel:

File: src/Core/Product/Search/SortOrderFactory.php [hunk 5]
```diff
@@ -28,33 +28,34 @@
...
-                $this->translator->trans('Name, Z to A', array(), 'Shop.Theme.Catalog')
-            ),
-            (new SortOrder('product', 'price', 'asc'))->setLabel(
-                $this->translator->trans('Price, low to high', array(), 'Shop.Theme.Catalog')
-            ),
-            (new SortOrder('product', 'price', 'desc'))->setLabel(
-                $this->translator->trans('Price, high to low', array(), 'Shop.Theme.Catalog')
-            ),
-        ];
+        return SortOrdersCollection::getDefaults();
```
Comment (mickaelandrieu) [inline src/Core/Product/Search/SortOrderFactory.php]: oups, it's not a static function, to be modified

**Review by eternoendless — DISMISSED** [review]
> Reviewed 11 of 19 files at r3, 2 of 3 files at r4, 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/prestashop/prestashop/10406)* status: all files reviewed, 5 unresolved discussions (waiting on @jolelievre and @mickaelandrieu)

**Comment by marionf** [comment]
@mickaelandrieu When I click on a category, all filters are selected

capture d ecran_1097

**Comment by marionf** [comment]
@mickaelandrieu 

I have a second page 1 with your PR

capture d ecran_1110

**Review by PierreRambaud — APPROVED** [review]

**Comment by PierreRambaud** [comment]
Thanks @mickaelandrieu

**Pull request merged**